### PR TITLE
add custom color scheme support

### PR DESCRIPTION
**Pull request by tussosedan**

### CODE REVIEW

**Comment by tao-yu** [comment]
Nice job! It seems to work well. 

Just one thing: putting something that isn't a colour into the the colourscheme boxes causes some strange behavior. The sim cube's colours act especially strange. Ideally, entering an invalid colour should warn the user. I see that you put that on your todo list, but I would like for this to be completed before I merge it because I don't want the weird behavior to be visible.

**Comment by tussosedan** [comment]
The todo is actually copied from your code for the rotations :)
No problem, I'll add a validation.

File: js/RubiksCube.js [hunk 1]
```diff
@@ -987,10 +1130,10 @@ var keymaps = [
 keymaps.forEach(function(keymap){
     listener.register_combo({
         "keys"              : keymap[0],
-        "on_keydown"        : function() {	doAlg(keymap[1]);},     
+        "on_keydown"        : function() {  doAlg(keymap[1]);},
```
Comment (tussosedan) [inline js/RubiksCube.js:1129]: side effect of converting all tabs to spaces

**Comment by tussosedan** [comment]
@tao-yu done! Did some refactoring of my changes too -- please take a look and see if all's well.

**Comment by tao-yu** [comment]
I'm afraid there's just a few more issues I'd like to see resolved before I merge this:

1. I would prefer if it was impossible for users to experience the weird sim cube behavior that results from invalid colors being entered. Sometimes, when you enter a invalid color, doing R on the sim cube can cause colors on the left side of the cube to change - this looks glitchy and don't think this is something users should see. Right now, if you enter an invalid color and press okay, you can still get this behavior (it does warn you every time you request a new scramble, but you can still play around with the cube at this point).

Here's the design I propose: when you enter a invalid color, it warns you. When you draw the cube, it draws the default color if it doesn't understand the color. The actual text in the box is unchanged - just in case the user simply made a typo. You keep getting warned until you fix the color or you reset the colors. 

2. I think it would be more elegant in this case to test for #RRGGBB codes using a regex:

`var isColor = /^#[0-9A-F]{6}$/i.test('#aabbcc')`

should work.

3. I think pressing the reset button should trigger a redraw of the virtual cube and also update visualcube. 

4. This is really a small issue, but it thinks "green " is an invalid color due to the space at the end. This doesn't seem to happen for color codes though. 

The first issue is really the main thing, but I would like the others to be done as well eventually. I may have time during the weekend to help out with some of these things if you'd like.

**Comment by tussosedan** [comment]
> 1. [...] When you draw the cube, it draws the default color if it doesn't understand the color. The actual text in the box is unchanged - just in case the user simply made a typo. You keep getting warned until you fix the color or you reset the colors.

Do you prefer this, or getting an error message and not generating a scramble?

> 2. I think it would be more elegant in this case to test for #RRGGBB codes using a regex:

At first I went for a generic color check that can handle both color names and hex -- but then it turned out that VisualCube only supports specific color names for some reason. The generic test can be useful in the future if this changes, or if you implement something else that can also receive color names.

Anyway I don't mind changing this if you prefer -- let me know (or just edit the pull request).

> 3. I think pressing the reset button should trigger a redraw of the virtual cube and also update visualcube.

You mean redraw the current cube state? I'm not sure it's consistent behavior, after all it's not updating while you're entering the custom color scheme. If you'd like that to happen though, I suppose we should also add an "apply" button to redraw the current state using the custom color scheme. What do you think?

> 4. This is really a small issue, but it thinks "green " is an invalid color due to the space at the end. This doesn't seem to happen for color codes though.

Should be easy enough to trim, I can do while working on the other points, after we decide on them.

**Comment by tao-yu** [comment]
> Do you prefer this, or getting an error message and not generating a scramble?

I actually like this idea better. When the user inputs invalid colors, make it give you an error when you press tab/enter, and leave the sim cube/visualcube untouched.

> At first I went for a generic color check that can handle both color names and hex -- but then it turned out that VisualCube only supports specific color names for some reason. The generic test can be useful in the future if this changes, or if you implement something else that can also receive color names.

Understood. We'll still have to check non hex code colors separately, and make sure they are compatible with visualcube. However, for codes that start with #, I think using a regex is clearer. We could also consider supporting three letter hex codes in the future. You can leave this part to me, I'll change it after the merge and see if I can support three letter codes as well. 

>You mean redraw the current cube state? I'm not sure it's consistent behavior, after all it's not updating while you're entering the custom color scheme. If you'd like that to happen though, I suppose we should also add an "apply" button to redraw the current state using the custom color scheme. What do you think?

I was playing around with qqtimer yesterday, which has a similar color picking interface to the one you designed, and if you change a color, the color gets updated as soon as you click out of the box. Maybe we could look into making this work? I think adding an onblur listener to each of the text boxes should work (an alert could also be added here, in addition to the one that pops up when pressing tab/enter). 

> Should be easy enough to trim, I can do while working on the other points, after we decide on them.

Great, as I said it's not a big issue. It's only possible when you copy and paste a space into the box, which I doubt many people will be doing, at least for non hex colors.

**Comment by tussosedan** [comment]
> When the user inputs invalid colors, make it give you an error when you press tab/enter, and leave the sim cube/visualcube untouched.

Thinking more about this, let's say an invalid color scheme is stored. What should happen when the page is first loaded -- should the cubes be displayed with default colors? Or no cubes displayed?

Also, this feature requires injecting checks into various places in the code in order to block the correct actions from happening, and not just the drawing of the cubes. I'm considering what's the most maintainable way to do it, hmm.

**Comment by tao-yu** [comment]
I think just setting invalid colors to their defaults is fine. I don't think there's a need to store invalid colors in localstorage. 

Everything looks good in this version. Can I ask you to just add this last thing and then I'll merge it? Thank you so much for all your work!

**Comment by tussosedan** [comment]
Sure, happy to help. The easiest solution would be to just reset the textbox to the default color and draw that. I agree it could be nice to keep the wrong text in case of a typo, but that means separating the displayed text and the colors actually drawn, which is a headache. Do you think resetting the textbox is good enough?

**Comment by tao-yu** [comment]
Yes, I think that's good enough. I don't think typos should be remembered between sessions.

**Comment by tussosedan** [comment]
Here we go, please take a look

**Comment by tao-yu** [comment]
I'm happy with this. Thank you for your help!

**Pull request merged**